File: Autonomous_Mobile_Robots_Course/Chapter9/Lesson3/Chapter9_Lesson3.py
# Chapter9_Lesson3.py
# Feature-Based Mapping (Landmarks) with a per-landmark EKF (pose assumed known).
# This script simulates a 2D robot trajectory, range-bearing observations to point landmarks,
# performs nearest-neighbor data association with Mahalanobis gating, initializes new landmarks,
# and updates landmark estimates with an EKF measurement update.

import numpy as np
import matplotlib.pyplot as plt

np.random.seed(7)

def wrap_angle(a):
    return (a + np.pi) % (2*np.pi) - np.pi

def h_rb(x, m):
    """Range-bearing measurement model z = [r, b] from pose x=[px,py,th] to landmark m=[mx,my]."""
    dx = m[0] - x[0]
    dy = m[1] - x[1]
    r = np.sqrt(dx*dx + dy*dy)
    b = wrap_angle(np.arctan2(dy, dx) - x[2])
    return np.array([r, b])

def H_rb_wrt_m(x, m):
    """Jacobian H = dh/dm for range-bearing wrt landmark position."""
    dx = m[0] - x[0]
    dy = m[1] - x[1]
    q = dx*dx + dy*dy
    r = np.sqrt(q)
    if r < 1e-9:
        r = 1e-9
        q = r*r
    # dr/dmx = dx/r , dr/dmy = dy/r
    # db/dmx = -dy/q , db/dmy = dx/q
    return np.array([[dx/r, dy/r],
                     [-dy/q,  dx/q]])

def inv_h_rb(x, z):
    """Inverse observation (used for initialization): landmark position from pose and measurement."""
    r, b = z
    ang = x[2] + b
    return np.array([x[0] + r*np.cos(ang), x[1] + r*np.sin(ang)])

def simulate_world():
    # True landmarks
    M_true = np.array([
        [-6.0,  5.0],
        [-2.0, -4.0],
        [ 4.0,  6.0],
        [ 7.0, -2.0],
        [ 1.0,  1.5],
        [-7.0, -6.0],
    ])
    # Robot trajectory (known poses for mapping): a lopsided loop
    T = 180
    t = np.linspace(0, 2*np.pi, T)
    px = 2.5*np.cos(t) + 0.5*np.cos(3*t)
    py = 2.0*np.sin(t)
    th = wrap_angle(np.arctan2(np.gradient(py), np.gradient(px)))
    X = np.vstack([px, py, th]).T

    # Sensor parameters
    r_max = 9.0
    sigma_r = 0.15
    sigma_b = np.deg2rad(2.0)
    R = np.diag([sigma_r**2, sigma_b**2])

    # Generate measurements (with unknown associations in the algorithm)
    Z = []  # list of list of (z, true_id)
    for k in range(T):
        zk = []
        for i, mi in enumerate(M_true):
            z = h_rb(X[k], mi)
            if z[0] <= r_max:
                z_noisy = z + np.array([np.random.normal(0, sigma_r),
                                        np.random.normal(0, sigma_b)])
                z_noisy[1] = wrap_angle(z_noisy[1])
                zk.append((z_noisy, i))
        Z.append(zk)
    return X, M_true, Z, R

class LandmarkMapEKF:
    def __init__(self, R, gate_chi2=9.21):
        """
        R: measurement covariance (2x2)
        gate_chi2: chi-square gate for 2 dof (e.g., 9.21 ~ 99%).
        """
        self.R = R
        self.gate = gate_chi2
        self.mu = []      # list of 2-vectors
        self.Sigma = []   # list of 2x2 covariances
        self.active = []  # list of bool
    def num(self):
        return len(self.mu)

    def add_landmark(self, x, z, init_sigma=1.5):
        m0 = inv_h_rb(x, z)
        self.mu.append(m0)
        self.Sigma.append(np.eye(2) * (init_sigma**2))
        self.active.append(True)

    def associate(self, x, z):
        """
        Nearest-neighbor association using Mahalanobis distance in measurement space.
        Returns (best_index or None, best_d2).
        """
        if self.num() == 0:
            return None, np.inf

        best_j, best_d2 = None, np.inf
        for j in range(self.num()):
            if not self.active[j]:
                continue
            m = self.mu[j]
            H = H_rb_wrt_m(x, m)
            zhat = h_rb(x, m)
            innov = np.array([z[0] - zhat[0], wrap_angle(z[1] - zhat[1])])
            S = H @ self.Sigma[j] @ H.T + self.R
            try:
                Sinv = np.linalg.inv(S)
            except np.linalg.LinAlgError:
                continue
            d2 = float(innov.T @ Sinv @ innov)
            if d2 < best_d2:
                best_d2 = d2
                best_j = j

        if best_d2 <= self.gate:
            return best_j, best_d2
        return None, best_d2

    def update(self, x, z, j):
        """EKF update of landmark j with measurement z at known pose x."""
        m = self.mu[j]
        P = self.Sigma[j]

        H = H_rb_wrt_m(x, m)
        zhat = h_rb(x, m)
        y = np.array([z[0] - zhat[0], wrap_angle(z[1] - zhat[1])])

        S = H @ P @ H.T + self.R
        K = P @ H.T @ np.linalg.inv(S)

        m_new = m + K @ y
        P_new = (np.eye(2) - K @ H) @ P

        self.mu[j] = m_new
        self.Sigma[j] = 0.5*(P_new + P_new.T)  # symmetrize

def run():
    X, M_true, Z, R = simulate_world()
    mapper = LandmarkMapEKF(R=R, gate_chi2=9.21)

    # Process measurements sequentially
    for k, x in enumerate(X):
        for (z, _true_id) in Z[k]:
            j, d2 = mapper.associate(x, z)
            if j is None:
                mapper.add_landmark(x, z)
            else:
                mapper.update(x, z, j)

    M_est = np.array(mapper.mu) if mapper.num() else np.zeros((0,2))

    # Plot
    plt.figure()
    plt.plot(X[:,0], X[:,1], '-', label="robot path (known)")
    plt.scatter(M_true[:,0], M_true[:,1], marker='x', s=80, label="true landmarks")
    if len(M_est):
        plt.scatter(M_est[:,0], M_est[:,1], marker='o', label="estimated landmarks")
        # draw covariance ellipses (2-sigma) for a subset
        for j in range(min(12, mapper.num())):
            P = mapper.Sigma[j]
            w, V = np.linalg.eigh(P)
            w = np.maximum(w, 1e-12)
            ang = np.linspace(0, 2*np.pi, 60)
            ell = (V @ (2.0*np.sqrt(w)[:,None]*np.vstack([np.cos(ang), np.sin(ang)])))
            plt.plot(mapper.mu[j][0] + ell[0,:], mapper.mu[j][1] + ell[1,:], linewidth=1)

    plt.axis('equal')
    plt.xlabel("x [m]")
    plt.ylabel("y [m]")
    plt.title("Feature-Based Mapping with per-landmark EKF")
    plt.legend()
    plt.grid(True)
    plt.show()

if __name__ == "__main__":
    run()

File: Tutorials/Codes/Autonomous_Mobile_Robots_Course/Chapter9/Lesson3/Chapter9_Lesson3.cpp
// Chapter9_Lesson3.cpp
// Feature-Based Mapping (Landmarks) with per-landmark EKF (pose assumed known).
// Dependencies: Eigen (header-only) for small matrix operations.
//
// Build (example):
//   g++ -O2 -std=c++17 Chapter9_Lesson3.cpp -I /path/to/eigen -o lesson3
// Run:
//   ./lesson3
//
// This program prints the final estimated landmark means (no plotting to keep dependencies minimal).

#include <Eigen/Dense>
#include <iostream>
#include <vector>
#include <random>
#include <cmath>
#include <limits>

static double wrap_angle(double a) {
  while (a > M_PI)  a -= 2.0*M_PI;
  while (a < -M_PI) a += 2.0*M_PI;
  return a;
}

struct Pose2 {
  double x, y, th;
};

struct MeasRB {
  double r, b; // range, bearing
};

static MeasRB h_rb(const Pose2& x, const Eigen::Vector2d& m) {
  double dx = m(0) - x.x;
  double dy = m(1) - x.y;
  double r = std::sqrt(dx*dx + dy*dy);
  double b = wrap_angle(std::atan2(dy, dx) - x.th);
  return {r, b};
}

static Eigen::Matrix2d H_rb_wrt_m(const Pose2& x, const Eigen::Vector2d& m) {
  double dx = m(0) - x.x;
  double dy = m(1) - x.y;
  double q  = dx*dx + dy*dy;
  double r  = std::sqrt(q);
  if (r < 1e-9) { r = 1e-9; q = r*r; }
  Eigen::Matrix2d H;
  H << dx/r, dy/r,
      -dy/q,  dx/q;
  return H;
}

static Eigen::Vector2d inv_h_rb(const Pose2& x, const MeasRB& z) {
  double ang = x.th + z.b;
  return Eigen::Vector2d(x.x + z.r*std::cos(ang), x.y + z.r*std::sin(ang));
}

struct Landmark {
  Eigen::Vector2d mu;
  Eigen::Matrix2d P;
  bool active;
};

struct Mapper {
  Eigen::Matrix2d R;
  double gate; // chi2 gate (2 dof)
  std::vector<Landmark> L;

  explicit Mapper(const Eigen::Matrix2d& R_in, double gate_chi2=9.21) : R(R_in), gate(gate_chi2) {}

  void add_landmark(const Pose2& x, const MeasRB& z) {
    Landmark lm;
    lm.mu = inv_h_rb(x, z);
    lm.P = Eigen::Matrix2d::Identity() * (1.5*1.5);
    lm.active = true;
    L.push_back(lm);
  }

  int associate(const Pose2& x, const MeasRB& z, double& best_d2) const {
    if (L.empty()) { best_d2 = std::numeric_limits<double>::infinity(); return -1; }
    int best_j = -1;
    best_d2 = std::numeric_limits<double>::infinity();

    for (int j=0; j<(int)L.size(); ++j) {
      if (!L[j].active) continue;
      const auto& lm = L[j];
      Eigen::Matrix2d H = H_rb_wrt_m(x, lm.mu);
      MeasRB zhat = h_rb(x, lm.mu);
      Eigen::Vector2d y(z.r - zhat.r, wrap_angle(z.b - zhat.b));
      Eigen::Matrix2d S = H*lm.P*H.transpose() + R;
      Eigen::Matrix2d Sinv = S.inverse();
      double d2 = y.transpose() * Sinv * y;
      if (d2 < best_d2) { best_d2 = d2; best_j = j; }
    }
    if (best_d2 <= gate) return best_j;
    return -1;
  }

  void update(const Pose2& x, const MeasRB& z, int j) {
    auto& lm = L[j];
    Eigen::Matrix2d H = H_rb_wrt_m(x, lm.mu);
    MeasRB zhat = h_rb(x, lm.mu);
    Eigen::Vector2d y(z.r - zhat.r, wrap_angle(z.b - zhat.b));
    Eigen::Matrix2d S = H*lm.P*H.transpose() + R;
    Eigen::Matrix2d K = lm.P * H.transpose() * S.inverse();
    lm.mu = lm.mu + K*y;
    lm.P  = (Eigen::Matrix2d::Identity() - K*H) * lm.P;
    lm.P  = 0.5*(lm.P + lm.P.transpose());
  }
};

int main() {
  std::mt19937 rng(7);
  std::normal_distribution<double> n01(0.0, 1.0);

  // True landmarks
  std::vector<Eigen::Vector2d> Mtrue = {
    {-6.0,  5.0}, {-2.0, -4.0}, { 4.0,  6.0},
    { 7.0, -2.0}, { 1.0,  1.5}, {-7.0, -6.0}
  };

  // Trajectory (known)
  const int T = 180;
  std::vector<Pose2> X; X.reserve(T);
  for (int k=0; k<T; ++k) {
    double t = 2.0*M_PI * (double)k/(double)(T-1);
    double px = 2.5*std::cos(t) + 0.5*std::cos(3.0*t);
    double py = 2.0*std::sin(t);
    // crude heading from derivative approximation
    double t2 = 2.0*M_PI * (double)std::min(k+1,T-1)/(double)(T-1);
    double px2 = 2.5*std::cos(t2) + 0.5*std::cos(3.0*t2);
    double py2 = 2.0*std::sin(t2);
    double th = wrap_angle(std::atan2(py2-py, px2-px));
    X.push_back({px, py, th});
  }

  // Sensor
  double r_max = 9.0;
  double sigma_r = 0.15;
  double sigma_b = 2.0 * M_PI / 180.0;
  Eigen::Matrix2d R;
  R << sigma_r*sigma_r, 0.0,
       0.0, sigma_b*sigma_b;

  Mapper mapper(R, 9.21);

  for (int k=0; k<T; ++k) {
    for (const auto& m : Mtrue) {
      MeasRB z = h_rb(X[k], m);
      if (z.r <= r_max) {
        // noisy measurement
        z.r += sigma_r * n01(rng);
        z.b = wrap_angle(z.b + sigma_b * n01(rng));

        double best_d2;
        int j = mapper.associate(X[k], z, best_d2);
        if (j < 0) mapper.add_landmark(X[k], z);
        else mapper.update(X[k], z, j);
      }
    }
  }

  std::cout << "Estimated landmarks (mu):\n";
  for (size_t j=0; j<mapper.L.size(); ++j) {
    std::cout << j << ": " << mapper.L[j].mu.transpose() << "\n";
  }
  return 0;
}
